src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +55,48 @@ public override void Initialize (AnalysisContext context)
 					}
 				}
 			});
-			// TODO: fix reporting for generic type substitutions. This should happen not only for method invocations,
-			// but for any reference to an instantiated method or type.
-			context.RegisterOperationAction (context => {
-				var invocationOperation = (IInvocationOperation) context.Operation;
-				ProcessInvocationOperation (context, invocationOperation);
-			}, OperationKind.Invocation);
+			context.RegisterSyntaxNodeAction (context => {
+				ProcessGenericParameters (context);
+			}, SyntaxKind.GenericName);
 		}
 
-		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		static void ProcessGenericParameters (SyntaxNodeAnalysisContext context)
 		{
-			if (context.ContainingSymbol.IsInRequiresUnreferencedCodeAttributeScope ())
+			// We dont need to check for RUC for INamedTypeSymbol here
+			if (context.ContainingSymbol is not null
+				&& context.ContainingSymbol is not INamedTypeSymbol
+				&& context.ContainingSymbol.IsInRequiresUnreferencedCodeAttributeScope ())
 				return;
 
-			ProcessTypeArguments (context, invocationOperation);
-		}
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).Type is INamedTypeSymbol type) {
+				// INamedTypeSymbol inside nameof, commonly used in [ExpectedWarning] can generate diagnostics
```

Comment:
nit: I would leave a comment about the expected behavior for `nameof` without mentioning `ExpectedWarning` - we should design the behavior for the customer code, and `ExpectedWarning` is just a detail of our codebase.